### ISSUE TITLE
API v1.0 review

### ISSUE THREAD

**Issue by pavelfeldman** [microsoft/playwright#1348]
### Playwright  ✅
- [x] playwright.chromium
- [x] playwright.devices
- [x] playwright.errors
- [x] playwright.firefox
- [x] playwright.selectors
- [x] playwright.webkit
- [x] Playwright module

### Browser ✅
- [x] event: 'disconnected'
- [x] browser.close()
- [x] browser.contexts()
- [x] browser.isConnected()
- [x] browser.newContext([options])
- [x] browser.newPage([options])

### BrowserContext ✅
- [x] event: 'close'
- [x] event: 'page'
- [x] browserContext.addInitScript(script[, arg])
- [x] browserContext.clearCookies()
- [x] browserContext.clearPermissions()
- [x] browserContext.close()
- [x] browserContext.cookies([urls])
- [x] browserContext.exposeFunction(name, playwrightFunction)
- [x] browserContext.newPage()
- [x] browserContext.pages()
- [x] browserContext.route(url, handler)
- [x] browserContext.addCookies(cookies)
- [x] browserContext.setDefaultNavigationTimeout(timeout)
- [x] browserContext.setDefaultTimeout(timeout)
- [x] browserContext.setExtraHTTPHeaders(headers)
- [x] browserContext.setGeolocation(geolocation)
- [x] browserContext.setHTTPCredentials(httpCredentials)
- [x] browserContext.setOffline(offline)
- [x] browserContext.grantPermissions(permissions[], options)
- [x] browserContext.waitForEvent(event[, optionsOrPredicate])

### BrowserType ✅
- [x] browserType.connect(options)
- [x] browserType.downloadBrowserIfNeeded([progress])
- [x] browserType.executablePath()
- [x] browserType.launch([options])
- [x] browserType.launchPersistentContext(userDataDir, [options])
- [x] browserType.launchServer([options])
- [x] browserType.name()

### BrowserServer ✅
- [x] event: 'close'
- [x] browserServer.close()
- [x] browserServer.kill()
- [x] browserServer.process()
- [x] browserServer.wsEndpoint()

### Worker ✅
- [x] event: 'close'
- [x] worker.evaluate(pageFunction[, arg])
- [x] worker.evaluateHandle(pageFunction[, arg])
- [x] worker.url()

### A11Y ✅
- [x] accessibility.snapshot([options])

### Selectors
- [x] selectors.register(name, script)

### Response ✅
- [x] response.body()
- [x] response.finished()
- [x] response.frame()
- [x] response.headers()
- [x] response.json()
- [x] response.ok()
- [x] response.request()
- [x] response.status()
- [x] response.statusText()
- [x] response.text()
- [x] response.url()

### Request ✅
- [x] request.abort([errorCode])
- [x] request.continue([overrides])
- [x] request.failure()
- [x] request.frame()
- [x] request.fulfill(response)
- [x] request.headers()
- [x] request.isNavigationRequest()
- [x] request.method()
- [x] request.postData()
- [x] request.redirectedFrom()
- [x] request.redirectedTo()
- [x] request.resourceType()
- [x] request.response()
- [x] request.url()

### Mouse ✅
- [x] mouse.click(x, y[, options])
- [x] mouse.dblclick(x, y[, options])
- [x] mouse.down([options])
- [x] mouse.move(x, y[, options])
- [x] mouse.up([options])

### Keyboard ✅
- [x] keyboard.down(key[, options])
- [x] keyboard.press(key[, options])
- [x] keyboard.insertText(text)
- [x] keyboard.type(text[, options])
- [x] keyboard.up(key)

### Dialog ✅
- [x] dialog.accept([promptText])
- [x] dialog.defaultValue()
- [x] dialog.dismiss()
- [x] dialog.message()
- [x] dialog.type()

### ConsoleMessage ✅
- [x] consoleMessage.args()
- [x] consoleMessage.location()
- [x] consoleMessage.text()
- [x] consoleMessage.type()

### JSHandle ✅
- [x] jsHandle.asElement()
- [x] jsHandle.dispose()
- [x] jsHandle.evaluate(pageFunction[, arg])
- [x] jsHandle.evaluateHandle(pageFunction[, arg])
- [x] jsHandle.getProperties()
- [x] jsHandle.getProperty(propertyName)
- [x] jsHandle.jsonValue()

### ElementHandle ✅
- [x] elementHandle.$(selector)
- [x] elementHandle.$$(selector)
- [x] elementHandle.$$eval(selector, pageFunction[, arg])
- [x] elementHandle.$eval(selector, pageFunction[, arg])
- [x] elementHandle.boundingBox()
- [x] elementHandle.check([options])
- [x] elementHandle.click([options])
- [x] elementHandle.contentFrame()
- [x] elementHandle.dblclick([options])
- [x] elementHandle.fill(value[, options])
- [x] elementHandle.focus()
- [x] elementHandle.hover([options])
- [x] elementHandle.ownerFrame()
- [x] elementHandle.press(key[, options])
- [x] elementHandle.screenshot([options])
- [x] elementHandle.scrollIntoViewIfNeeded()
- [x] elementHandle.selectOption(values[, options])
- [x] elementHandle.setInputFiles(files)
- [x] elementHandle.toString()
- [x] elementHandle.type(text[, options])
- [x] elementHandle.uncheck([options])

### Frame ✅
- [x] frame.$(selector)
- [x] frame.$$(selector)
- [x] frame.$$eval(selector, pageFunction[, arg])
- [x] frame.$eval(selector, pageFunction[, arg])
- [x] frame.addScriptTag(options)
- [x] frame.addStyleTag(options)
- [x] frame.check(selector, [options])
- [x] frame.childFrames()
- [x] frame.click(selector[, options])
- [x] frame.content()
- [x] frame.dblclick(selector[, options])
- [x] frame.evaluate(pageFunction[, arg])
- [x] frame.evaluateHandle(pageFunction[, arg])
- [x] frame.fill(selector, value[, options])
- [x] frame.focus(selector[, options])
- [x] frame.frameElement()
- [x] frame.goto(url[, options])
- [x] frame.hover(selector[, options])
- [x] frame.isDetached()
- [x] frame.name()
- [x] frame.parentFrame()
- [x] frame.press(selector, key[, options])
- [x] frame.selectOption(selector, values[, options])
- [x] frame.setContent(html[, options])
- [x] frame.title()
- [x] frame.type(selector, text[, options])
- [x] frame.uncheck(selector, [options])
- [x] frame.url()
- [x] frame.waitFor(selectorOrFunctionOrTimeout[, options[, arg]])
- [x] frame.waitForFunction(pageFunction[, options[, arg]])
- [x] frame.waitForLoadState([options])
- [x] frame.waitForNavigation([options])
- [x] frame.waitForSelector(selector[, options])

### Page ✅
- [x] event: 'close'
- [x] event: 'console'
- [x] event: 'dialog'
- [x] event: 'domcontentloaded'
- [x] event: 'filechooser'
- [x] event: 'frameattached'
- [x] event: 'framedetached'
- [x] event: 'framenavigated'
- [x] event: 'load'
- [x] event: 'pageerror'
- [x] event: 'popup'
- [x] event: 'request'
- [x] event: 'requestfailed'
- [x] event: 'requestfinished'
- [x] event: 'response'
- [x] event: 'worker'
- [x] page.$(selector)
- [x] page.$$(selector)
- [x] page.$$eval(selector, pageFunction[, arg])
- [x] page.$eval(selector, pageFunction[, arg])
- [x] page.accessibility
- [x] page.addInitScript(script[, arg])
- [x] page.addScriptTag(options)
- [x] page.addStyleTag(options)
- [x] page.check(selector, [options])
- [x] page.click(selector[, options])
- [x] page.close([options])
- [x] page.content()
- [x] page.context()
- [x] page.coverage
- [x] page.dblclick(selector[, options])
- [x] page.emulateMedia(options)
- [x] page.evaluate(pageFunction[, arg])
- [x] page.evaluateHandle(pageFunction[, arg])
- [x] page.exposeFunction(name, playwrightFunction)
- [x] page.fill(selector, value[, options])
- [x] page.focus(selector[, options])
- [x] page.frame(options)
- [x] page.frames()
- [x] page.goBack([options])
- [x] page.goForward([options])
- [x] page.goto(url[, options])
- [x] page.hover(selector[, options])
- [x] page.isClosed()
- [x] page.keyboard
- [x] page.mainFrame()
- [x] page.mouse
- [x] page.opener()
- [x] page.pdf([options])
- [x] page.press(selector, key[, options])
- [x] page.reload([options])
- [x] page.route(url, handler)
- [x] page.screenshot([options])
- [x] page.selectOption(selector, values[, options])
- [x] page.setContent(html[, options])
- [x] page.setDefaultNavigationTimeout(timeout)
- [x] page.setDefaultTimeout(timeout)
- [x] page.setExtraHTTPHeaders(headers)
- [x] page.setViewportSize(viewportSize)
- [x] page.title()
- [x] page.type(selector, text[, options])
- [x] page.uncheck(selector, [options])
- [x] page.url()
- [x] page.viewportSize()
- [x] page.waitFor(selectorOrFunctionOrTimeout[, options[, arg]])
- [x] page.waitForEvent(event[, optionsOrPredicate])
- [x] page.waitForFunction(pageFunction[, options[, arg]])
- [x] page.waitForLoadState([options]) 
- [x] page.waitForNavigation([options])
- [x] page.waitForRequest(urlOrPredicate[, options])
- [x] page.waitForResponse(urlOrPredicate[, options])
- [x] page.waitForSelector(selector[, options])
- [x] page.workers()

### CDPSession ✅
- [x] cdpSession.detach()
- [x] cdpSession.send(method[, params])

### ChromiumCoverage ✅
- [x] chromiumCoverage.startCSSCoverage([options])
- [x] chromiumCoverage.startJSCoverage([options])
- [x] chromiumCoverage.stopCSSCoverage()
- [x] chromiumCoverage.stopJSCoverage()

### ChromiumBrowserContext✅
- [x] event: 'backgroundpage'
- [x] event: 'serviceworker'
- [x] chromiumBrowserContext.backgroundPages()
- [x] chromiumBrowserContext.serviceWorkers()
- [x] chromiumBrowserContext.newCDPSession(page)

### ChromiumBrowser ✅
- [x] chromiumBrowser.newBrowserCDPSession()
- [x] chromiumBrowser.startTracing([page, options])
- [x] chromiumBrowser.stopTracing()

**Comment by JoelEinbinder** [comment]
> browserContext.setDefaultNavigationTimeout, browserContext.setDefaultTimeout will change to browserContext.setDefaultTimeouts({ navigation, click, type })
Rationale: There are numerous delays for clicks and types and what not, would be nice to configure in a generic manner

Sure there are different places delays are used, but what about each thing let’s me decide what the delay should be? Trying to tune individual timeouts seems like an “always wrong” strategy to me.

**Comment by JoelEinbinder** [comment]
> browserContext.waitForEvent(event[, optionsOrPredicate])

This looks like we are fighting the node standard library. The old waitForTarget made sense because it included existing targets. If waitForEvent doesn’t do that, I’d rather just remove it. Why don’t we have it on workers?

**Comment by JoelEinbinder** [comment]
> browserType.downloadBrowserIfNeeded([progress])

Due to yarn, I believe this needs to take a download folder. @aslushnikov should remember the details.

**Comment by JoelEinbinder** [comment]
> browserType.errors

Let’s remove the shortcuts to errors on the browserTypes. We aren’t saving anyone anything in practice. TypeScript users need to import playwright for types already. I think other users are more likely to get passed a browser/context than a browser type. It feels out of place.

**Comment by JoelEinbinder** [comment]
> browserServer.close()
browserServer.kill()

When do I choose kill? Maybe we know based on how the server was launched (persistent)? If kill isn’t recommended, maybe we can let people do it through the process if they need to.

**Comment by JoelEinbinder** [comment]
s/sendCharacters/insertText

**Comment by aslushnikov** [comment]
> When do I choose kill? Maybe we know based on how the server was launched (persistent)? If kill isn’t recommended, maybe we can let people do it through the process if they need to.

@JoelEinbinder killing is involved across OS'es (e.g. on win we use `taskkill`). Also, killing can be called from sync event handlers, e.g. `process.on('SIGTERM')` handler. These are probably handy for embedding testrunners.

**Comment by JoelEinbinder** [comment]
> > When do I choose kill? Maybe we know based on how the server was launched (persistent)? If kill isn’t recommended, maybe we can let people do it through the process if they need to.
> 
> @JoelEinbinder killing is involved across OS'es (e.g. on win we use `taskkill`). Also, killing can be called from sync event handlers, e.g. `process.on('SIGTERM')` handler. These are probably handy for embedding testrunners.

So if I am in SIGTERM I call kill, otherwise I call close? If I’m in sigterm and I don’t do anything, will we leak processes?

**Comment by aslushnikov** [comment]
> So if I am in SIGTERM I call kill, otherwise I call close?

@JoelEinbinder the comparison table between the two:
- `browser.kill()` is a process kiling. **PROs:** it's fast & sync. **CONs:** doesn't guarantee to save any browser data and might generate a crashdump if your OS is configured to do so.
- `browser.close()` is a graceful shutdown. **PROs:** saves your browser data and doesn't generate crashdump. **CONs**: slow & non-sync.

>  If I’m in sigterm and I don’t do anything, will we leak processes?

By default, PW handles `SIGINT`/`SIGTERM`/`SIGHUP`  and closes browser appropriately. This behavior is configurable with launch flags.

If parent node.js process terminates, Playwright will kill the launched browser.
